primitives/src/pcs/univariate_kzg/mod.rs

```diff
@@ -652,6 +655,81 @@ where
     }
 }
 
+#[cfg(feature = "icicle")]
+impl<E: Pairing> UnivariateKzgPCS<E> {
+    /// Comupte PCS commit using GPU
+    /// Similar to [`Self::commit()`] but with ICICE's GPU-accelerated MSM
+    pub fn commit_with_gpu<C>(
+        prover_param: impl Borrow<UnivariateProverParam<E>>,
+        poly: &DensePolynomial<E::ScalarField>,
+    ) -> Result<Commitment<E>, PCSError>
+    where
+        C: IcicleCurve + MSM<C>,
+        C::ScalarField: ArkConvertible<ArkEquivalent = E::ScalarField>,
+        C::BaseField: ArkConvertible<ArkEquivalent = <C::ArkSWConfig as CurveConfig>::BaseField>,
+        E: Pairing<G1Affine = Affine<<C as IcicleCurve>::ArkSWConfig>>,
+    {
+        // TODO: use proper PCSError instead of unwrap.
+        #[cfg(feature = "kzg-print-trace")]
+        let commit_time =
+            start_timer!(|| format!("Committing to polynomial of degree {} ", poly.degree()));
+
+        let prover_param = prover_param.borrow();
+        if poly.degree() > prover_param.powers_of_g.len() {
+            return Err(PCSError::InvalidParameters(format!(
+                "poly degree {} is larger than allowed {}",
+                poly.degree(),
+                prover_param.powers_of_g.len()
+            )));
+        }
+
+        #[cfg(feature = "kzg-print-trace")]
+        let conv_time = start_timer!(|| "Type Conversion: arkworks->ICICLE");
+
+        let size = poly.degree() + 1;
+        let bases: Vec<IcicleAffine<C>> = prover_param.powers_of_g[..size]
```

Comment:
Did some profilings on my local computer with degree 2^20, this type conversion consumes quite a lot of time.
Consider using `parallelizable_slice_iter` instead.

```
Start:   KZG10::Setup with prover degree 1048576 and verifier degree 1
··Start:   Generating powers of G
··End:     Generating powers of G ..................................................637.759ms
End:     KZG10::Setup with prover degree 1048576 and verifier degree 1 .............695.933ms
Start:   Committing to polynomial of degree 1048576
··Start:   Type Conversion: arkworks->ICICLE
··End:     Type Conversion: arkworks->ICICLE .......................................382.334ms
··Start:   GPU-accelerated MSM
··End:     GPU-accelerated MSM .....................................................38.753ms
End:     Committing to polynomial of degree 1048576  ...............................730.142ms
Start:   Opening polynomial of degree 1048576
··Start:   Computing witness polynomial
··End:     Computing witness polynomial ............................................92.564ms
··Start:   Converting polynomial coeffs to
    bigints
··End:     Converting polynomial coeffs to
    bigints .............................12.859ms
End:     Opening polynomial of degree 1048576 ......................................932.008ms
Start:   Checking evaluation
End:     Checking evaluation Result: true...........................................1.059ms
```

Reply:
And the bench result after using parallelizable_slice_iter
```
MSM with ICICLE/19      time:   [42.931 ms 43.461 ms 43.881 ms]
MSM with ICICLE/20      time:   [76.738 ms 77.138 ms 77.910 ms]
MSM with ICICLE/21      time:   [143.75 ms 145.62 ms 148.23 ms]
MSM with ICICLE/22      time:   [272.33 ms 275.70 ms 279.48 ms]
```

Reply:
Their type conversion is also not optimal, see
https://github.com/ingonyama-zk/icicle/blob/e6035698b5e54632f2c44e600391352ccc11cad4/wrappers/rust/icicle-core/src/field.rs#L68-L82
and
https://github.com/ingonyama-zk/icicle/blob/e6035698b5e54632f2c44e600391352ccc11cad4/wrappers/rust/icicle-core/src/field.rs#L125-L135

We can at least avoid one memory copy here.

Reply:
Moreover, sad that the conversion is not trivial.
Looks like icicle stores the field elements as in plain u64 limbs, while artworks stores them in Montgomery form.
But it's not complicated, definitely we do not need those memory copies.

Reply:
yah those "conversion dominated" breakdown makes lots of sense. thanks for helping on this! 

I'll improve as you suggested!

Reply:
`par_iter` added in [250e6f4](https://github.com/EspressoSystems/jellyfish/pull/498/commits/250e6f450db94ab2b232b6d10b6abe678dca83f7), 

>Their type conversion is also not optimal

i assume you are saying that they don't need to convert to bytes then back, instead, they can just use bigint?
if so, it's the upstream that needs to be changed. plus the memory copy should be neglibile and fast

Reply:
> ^^ somehow it feels like my par_iter is not working, the conversion part still takes a long time
(I suspect it's because my CPU core is limited only 4 cores, 8 threads 🤔

Yes, it heavily depends on number of cores.

> i assume you are saying that they don't need to convert to bytes then back, instead, they can just use bigint?
if so, it's the upstream that needs to be changed. plus the memory copy should be neglibile and fast

Sure we could nail a PR in upstream repos. But I think we could have our own converter.